### ISSUE TITLE
MongoDB Atlas reports too many connections to the database

### ISSUE THREAD

**Issue by psonnera** [nightscout/cgm-remote-monitor#8024]
**If you need support for Nightscout, PLEASE DO NOT FILE A TICKET HERE**
For support, please post a question to the "CGM in The Cloud" group in Facebook
(https://www.facebook.com/groups/cgminthecloud) or visit the WeAreNotWaiting Discord at https://discord.gg/zg7CvCQ

**Describe the bug**
We have several reports of AAPS/Loop users receiving a mail from MongoDB specifying they are using more than 500 connections and M0 doesn't support this. As a result, Nightscout will crash.

**To Reproduce**
Not known yet, users confirm they didn't change the amount of uploaders/downloaders

**Expected behavior**
500 connections seem impossible for Nightscout

**Screenshots**

![image](https://github.com/nightscout/cgm-remote-monitor/assets/33809361/d078b7ca-ca00-4f17-b9fd-033d5922da99)
![image](https://github.com/nightscout/cgm-remote-monitor/assets/33809361/d45b1fc2-a64c-4564-963c-96e21e9096f4)

**Your setup information**
- Nightscout 14.2.x
- AAPS/Loop

**Additional context**
Not known yet

(Will edit and update with new information, not using Atlas I haven't seen the issue)

**Comment by psonnera** [comment]
The same issue has been seen with Railway
![image](https://github.com/nightscout/cgm-remote-monitor/assets/33809361/cae440dd-f38d-4a79-ae60-63435063d28a)

**Comment by sulkaharo** [comment]
This is a pretty bizarre issue. Nightscout doesn't do any connection management, with the assumption that the MongoDB driver pools the connections and would never need more than a few concurrent connections given we're reusing the connection object. This makes me wonder if there's been an update to the driver that requires code changes, there's a bug in the MongoDB driver or if there's a bug in Nightscout that causes runaway connections / connection leaks. The relevant code in NS hasn't changed and I never observed that many connections being used. @bewest ideas?

**Comment by bewest** [comment]
Similar outlook, @sulkaharo.  Maybe the MongoDB version?  Are they using Mongo 6 or some version that changes the backwards compatibility?

**Comment by bewest** [comment]
What is the time period for these connections and how does Atlas measure them?  It's possible that if the server is shedding connections, the client will have to reconnect.  It's also possible that if Nightscout is crashing, perhaps due to an AAPS data issue, that restarting lots of times will burn through more connections.  It's worth looking carefully to see if the crashes are causing the connections to increase vs other way around.  In general in the case of a db error, Nightscout should show a detailed error, page not crash.  However, crashing will generate 8 new connections when the hosting provider restarts the process.  What's in the crash logs?

**Comment by jamesthurlow** [comment]
@bewest version 6 I think.  Just trying to figure out how to get access to crash logs.

![image](https://github.com/nightscout/cgm-remote-monitor/assets/16494970/52568e6f-a306-4185-a2ea-03d65714ed12)

**Comment by sulkaharo** [comment]
Isn't this concurrent connections though? If so, Nightscout crashing can't be the culprit as crashing should release the connections. I recently observed my instance having gone to a mode where some event that causes the data reload was being triggered very often - if that happens in a runaway way, that could cause a lot of connections being used. I'll add a constraint to how often that can happen.

**Comment by sulkaharo** [comment]
Right, so bootevent already implements debounce, but it's only set to cap the data updates to once / 5 seconds, which I suspect is too fast for Atlas. Depending on how they defer the execution of queries, I guess this could cause 500 connections to be consumed if something was causing the data update event to be triggered frequently. I suggest we raise that DB load debounce to 15 seconds and see if that helps. With the current server implementation, this load is theoretically not needed at all.

The other potential culprit is the ddata_at REST endpoint, which has no cap to how frequently it can be called and it'd be very easy to take an instance down by calling that at a rapid rate.

**Comment by bewest** [comment]
@jamesthurlow, if your Nightscout is hosted in Heroku, then it will be there, or Railway, Flyio, etc.
Before this issue, were you previously hitting the Atlas size quotas and then deleted some data?
Maybe we can add some instrumentation to the pool to see if we can monitor the issue ourselves?

**Comment by ninelore** [comment]
Hello everyone, this issues recently appeared after I upgraded to 15.0.2 on fly.io.

Connections went down in Atlas after i shut down my nightscout instance for testing 
![image](https://github.com/nightscout/cgm-remote-monitor/assets/21343557/32ec7f6f-018d-4409-8acf-a301765e09be)

**Comment by AndyLow91** [comment]
I’m going to close this now as superseded by later MongoDB connection-pool changes.

This was originally reported against Nightscout `14.2.x` / `15.0.2`, where the MongoDB driver could maintain a much larger default connection pool. Current Nightscout now has explicit MongoDB pool controls and a much smaller default pool size:

- `MONGO_POOL_SIZE` defaults to `5`
- `MONGO_MIN_POOL_SIZE` defaults to `0`
- `MONGO_MAX_IDLE_TIME_MS` defaults to `30000`
- optional pool debug logging is available for investigation

That should materially reduce the risk of a single Nightscout process consuming hundreds of Atlas connections, and it also gives maintainers/users a clear way to tune or inspect the pool if a deployment needs it.

If anyone still sees Atlas reporting excessive connections on a current Nightscout version, please open a fresh issue with:

- current Nightscout version/image tag
- hosting provider and instance count
- MongoDB Atlas tier
- configured `MONGO_POOL_SIZE`, `MONGO_MIN_POOL_SIZE`, and `MONGO_MAX_IDLE_TIME_MS`
- whether the app is restarting repeatedly
- Atlas connection graph and Nightscout logs from the same time window

That should make it easier to determine whether there is still a current connection leak, a restart loop, multiple running app instances, or Atlas-side connection accounting.
